pkg/operators/combiner/combiner.go

```diff
@@ -0,0 +1,309 @@
+				combinedDs.Name(), err)
+		}
+
+		// Allocate a new packet array for next iteration
+		combinedPacket, err = combinedDs.NewPacketArray()
+		if err != nil {
+			// We can't continue without a new packet array
+			return fmt.Errorf("creating new packet array: %w", err)
+		}
+
+		targetCount = 0
+		return nil
+	}
+
+	for {
+		select {
+		case <-o.done:
+			gadgetCtx.Logger().Debug("combiner: done")
+			return
+		case inPacket := <-packetBuf:
+			if inPacket == nil {
+				continue
+			}
+
+			targetCount++
+			gadgetCtx.Logger().Debugf("combiner: received data from %d/%d targets", targetCount, o.targets)
+
+			// TODO: Support Append multiple packets at once?
+			for i := 0; i < inPacket.Len(); i++ {
+				combinedPacket.Append(inPacket.Get(i))
+			}
+
+			// For data sources that doesn't have an interval, we wait for data
+			// from all targets before emitting the combined data.
+			if interval == 0 && targetCount == o.targets {
+				if err := emitAndAllocate(); err != nil {
+					gadgetCtx.Logger().Errorf("emitting and allocating data: %s", err)
+				}
+				return
+			}
+		case <-c:
+			if interval == 0 {
+				gadgetCtx.Logger().Warnf("Data is incomplete: timeout waiting for data from all targets (%d/%d)",
+					targetCount, o.targets)
+			}
+
+			if err := emitAndAllocate(); err != nil {
+				gadgetCtx.Logger().Errorf("emitting and allocating data: %s", err)
+				return
+			}
+		}
+	}
+}
+
+func (o *combinerOperatorInstance) PreStart(gadgetCtx operators.GadgetContext) error {
+	o.done = make(chan struct{})
+
+	for _, ds := range gadgetCtx.GetDataSources() {
+		if ds.Type() != datasource.TypeArray {
+			continue
+		}
+
+		interval, err := getFetchAnnotation(ds)
+		if err != nil {
+			return fmt.Errorf("getting fetch annotation for ds %s: %w", ds.Name(), err)
+		}
+
+		// Disable original data source to avoid other operators subscribing to it
+		ds.SetRequested(false)
+
+		// Register a new data source that will emit the combined data
+		combinedDs, err := gadgetCtx.RegisterDataSource(
+			datasource.TypeArray,
+			fmt.Sprintf("%s-%s", DataSourcePrefix, ds.Name()),
+		)
+		if err != nil {
+			return fmt.Errorf("registering combiner data source for %s: %w", combinedDs.Name(), err)
+		}
+
+		gadgetCtx.Logger().Debugf("%s: registered ds %s", o.Name(), combinedDs.Name())
+
+		// Use the same fields and annotations as the original data source
+		ds.CopyFieldsTo(combinedDs)
+		for k, v := range ds.Annotations() {
+			combinedDs.AddAnnotation(k, v)
+		}
+
+		o.pktBuffers[ds.Name()] = make(chan datasource.PacketArray, o.targets)
+		go o.forwardData(gadgetCtx, ds.Name(), interval, combinedDs)
+
+		gadgetCtx.Logger().Debugf("%s: subscribing to %s", o.Name(), ds.Name())
+
+		ds.SubscribePacket(func(source datasource.DataSource, packet datasource.Packet) error {
+			gadgetCtx.Logger().Debug("Received data array ... Sending to combiner")
+
+			// Use Marshal/Unmarshal to create a deep copy
+			b, _ := proto.Marshal(packet.Raw())
```

Comment:
Is `proto.Clone()` an option here? If it can't be used directly, maybe it makes sense to make that a member of `datasource.Packet`.

Reply:
To make it be:

```go
			// Deep copy and send it to the combiner data source
			clonedPkt := packet.Clone()
			pktBuffers <- clonedPkt
```

Reply:
Yeah, looks good to me. I don't have numbers, but I would guess this is also faster.

Reply:
I'll add it to the TODO list for follow-up optimizations.